### PR TITLE
Rework SplitButton to not use SWT's internal TypedListener

### PR DESCRIPTION
**Pull request by HannesWell**
The `org.eclipse.swt.widgets.TypedListener` should be considered an internal class (as the doc states) and leaks the class `SWTEventListener` which also resides in an internal package.

This reworks `SplitButton` to not use `TypedListener` anymore.

### CODE REVIEW

File: widgets/splitbutton/org.eclipse.nebula.widgets.splitbutton/src/org/eclipse/nebula/widgets/splitbutton/SplitButton.java [hunk 1]
```diff
@@ -179,30 +179,29 @@ private void drawTriangle() {
...
+			} else {
+				Supplier<Event> eventFactory = () -> {
+					Event evt = new Event();
+					evt.widget = SplitButton.this;
+					evt.display = getDisplay();
+					evt.type = SWT.Selection;
+					return evt;
+				};
+				for (Listener listener : selectionListeners) {
+					listener.handleEvent(eventFactory.get());
```
Comment (laeubi) [inline widgets/splitbutton/org.eclipse.nebula.widgets.splitbutton/src/org/eclipse/nebula/widgets/splitbutton/SplitButton.java]: Actually it looks wrong to create a new event for each listener, @lcaron what do you think?

**Comment by merks** [comment]
Unless bounds are changed that will allow this latest Nebula bundle to only installed along with latest version of SWT, this type of change is going to cause grief.

If it were "my project" I would probably try to create a utility that allowed me to work reflectively with both the old "APIs" and the new APIs so that these things continue to "just work".

File: widgets/splitbutton/org.eclipse.nebula.widgets.splitbutton/src/org/eclipse/nebula/widgets/splitbutton/SplitButton.java (hunk 1 above)
Comment (HannesWell) [reply widgets/splitbutton/org.eclipse.nebula.widgets.splitbutton/src/org/eclipse/nebula/widgets/splitbutton/SplitButton.java]: It used to be like that, this is why I replicated it.
But when looking at SWT I get the impression that there is only one Event send to all listeners, unless one uses a typed listener (e.g. SelectionListener), then the one event is wrapped in a new `SelectionEvent` for each `SelectionListener`, see `TypedListener`. But the latter is probably a technical detail.
And although Events are not immutable, I think good listeners don't change events.

**Comment by HannesWell** [comment]
> Unless bounds are changed that will allow this latest Nebula bundle to only installed along with latest version of SWT, this type of change is going to cause grief.

This change does not use the recent API addition in SWT (as you can see in the CI it builds with the latest SimRel) and just avoids `TypedListener` by reworking the listener handling. I assume that's the source of your concern, isn't it?
Although I don't know if this is compatible with SWT 1.0, but it shouldn't raise the requirements on the SWT version compared to the previous state.

> If it were "my project" I would probably try to create a utility that allowed me to work reflectively with both the old "APIs" and the new APIs so that these things continue to "just work".

The new APIs are only used in https://github.com/eclipse/nebula/pull/583, but there I raised the version bounds. But lets discuss there if any other strategy should be used.

**Comment by merks** [comment]
Sorry, I didn't look in detail of what's being used!

File: widgets/splitbutton/org.eclipse.nebula.widgets.splitbutton/src/org/eclipse/nebula/widgets/splitbutton/SplitButton.java (hunk 1 above)
Comment (merks) [reply widgets/splitbutton/org.eclipse.nebula.widgets.splitbutton/src/org/eclipse/nebula/widgets/splitbutton/SplitButton.java]: Except for "doit" when the event has that sort of thing.

Comment (HannesWell) [reply widgets/splitbutton/org.eclipse.nebula.widgets.splitbutton/src/org/eclipse/nebula/widgets/splitbutton/SplitButton.java]: > Except for "doit" when the event has that sort of thing.

Didn't knew that, thanks for the hint. So maybe there should be one `Event` send to all `Listeners` and a new `SelectionEvent` for each `SelectionListener`?

Comment (merks) [reply widgets/splitbutton/org.eclipse.nebula.widgets.splitbutton/src/org/eclipse/nebula/widgets/splitbutton/SplitButton.java]: I think that's how SWT generally does it...

Comment (laeubi) [reply widgets/splitbutton/org.eclipse.nebula.widgets.splitbutton/src/org/eclipse/nebula/widgets/splitbutton/SplitButton.java]: > And although Events are not immutable, I think good listeners don't change events.

Actually listeners can give some things back to the caller (e.g. as mentioned doit), there is even a field `data` mentioned for "application use" in TypedEvent, I think if one can install listeners to an SWT object one should always be "nice" and only ever modify things if the API tells so.

**Comment by HannesWell** [comment]
@lcaron, @wimjongman can you please review this as well?
This only uses already existing SWT API.

File: widgets/splitbutton/org.eclipse.nebula.widgets.splitbutton/src/org/eclipse/nebula/widgets/splitbutton/SplitButton.java [hunk 2]
```diff
@@ -179,30 +179,26 @@ private void drawTriangle() {
...
-				} else {
-					for (final Listener listener : listeners) {
-						final Event evt = new Event();
-						evt.widget = SplitButton.this;
-						evt.display = getDisplay();
-						evt.type = SWT.Selection;
-						listener.handleEvent(evt);
-					}
+		super.addListener(SWT.MouseDown, event -> {
+			boolean shouldShowMenu = event.x >= splitButtonAreaLeft && event.y >= splitButtonAreaTop
```
Comment (wimjongman) [inline widgets/splitbutton/org.eclipse.nebula.widgets.splitbutton/src/org/eclipse/nebula/widgets/splitbutton/SplitButton.java]: For readability the shouldShowMenu method should not be inlined.

**Review by wimjongman — COMMENTED** [review]
One remark. Otherwise LGTM. I did not run it so I don't know if the widget works the same way it used to.

File: widgets/splitbutton/org.eclipse.nebula.widgets.splitbutton/src/org/eclipse/nebula/widgets/splitbutton/SplitButton.java (hunk 2 above)
Comment (HannesWell) [reply widgets/splitbutton/org.eclipse.nebula.widgets.splitbutton/src/org/eclipse/nebula/widgets/splitbutton/SplitButton.java]: Ok, just reverted that part.

**Comment by HannesWell** [comment]
> One remark. Otherwise LGTM. I did not run it so I don't know if the widget works the same way it used to.

I'm not familiar with Nebula in general and the `SplitButton` especially, but when I run the `SplitButtonSnippet` everything seem to work fine as far as I understood it.

**Comment by HannesWell** [comment]
@wimjongman is there anything missing here or can this be submitted?

**Comment by wimjongman** [comment]
Go for it. Thanks, Hannes!

**Comment by HannesWell** [comment]
> Go for it. Thanks, Hannes!

I'm not a committer here, but thanks Ed for submitting.

**Comment by wimjongman** [comment]
> > Go for it. Thanks, Hannes!
> 
> I'm not a committer here, but thanks Ed for submitting.

You get a special mention when we release. 🏅

**Pull request merged**